### PR TITLE
updating comment based help and platyPS tasks to run out of process (…

### PR DESCRIPTION
**Pull request by gaelcolas**
Thanks to @pshamus

Here's the draft I'm working on.

I'm not sure about the workflow yet:
- if we have Comment based help, what should we do?
	- Generate markdown help files in Source?
	- Generate the markdown help files in output and build Maml from there (it won't be clean)
- if we only have Markdown help, just generate the Maml and ship it in en-US or whatever folder?

Let me knwo how you see the workflow so I can finish this up.

---
This change is [Reviewable](https://reviewable.io/reviews/gaelcolas/sampler/260)

### CODE REVIEW

File: .build/tasks/generateHelp.PlatyPS.build.ps1 [hunk 1]
```diff
@@ -41,26 +48,162 @@ Task UpdateHelp {
...
+    $HelpOutputVersionFolder = Get-SamplerAbsolutePath -Path $ModuleVersion -RelativeTo $HelpOutputFolder
+    "`tHelp output Version Folder= '$HelpOutputVersionFolder'"
+
+    $HelpOutputCultureFolder = Get-SamplerAbsolutePath -Path $HelpCultureInfo -RelativeTo $HelpOutputVersionFolder
+    "`tHelp output Culture path  = '$HelpOutputCultureFolder'"
+
+    $generateHelpCommands = @(
+        '$env:PSModulePath = "{0}"' -f $Env:PSModulePath
+        'Import-Module -Name "{0}" -Force' -f $ProjectName
+        '$newMarkdownHelpParams = @{{ Module = "{0}"; OutputFolder = "{1}"; AlphabeticParamsOrder = ${2}; WithModulePage = ${3}; ExcludeDontShow = ${4}; Force = $true}}' -f $ProjectName, $HelpOutputCultureFolder, $true, $true, $true
```
Comment (pshamus) [inline .build/tasks/generateHelp.PlatyPS.build.ps1:116]: Maybe make `AlphabeticParamsOrder` a variable with a default value that can be overridden. I usually set that to `$false`.

File: .build/tasks/generateHelp.PlatyPS.build.ps1 [hunk 2]
```diff
@@ -41,26 +48,162 @@ Task UpdateHelp {
...
+    "`tHelp Folder Path         = '$HelpFolder'"
+
+    $HelpOutputFolder =  Get-SamplerAbsolutePath -Path $HelpOutputFolder -RelativeTo $OutputDirectory
+    "`tHelp output Folder        = '$HelpOutputFolder'"
+
+    $HelpOutputVersionFolder = Get-SamplerAbsolutePath -Path $ModuleVersion -RelativeTo $HelpOutputFolder
+    "`tHelp output Version Folder= '$HelpOutputVersionFolder'"
+
+    $HelpOutputCultureFolder = Get-SamplerAbsolutePath -Path $HelpCultureInfo -RelativeTo $HelpOutputVersionFolder
+    "`tHelp output Culture path  = '$HelpOutputCultureFolder'"
```
Comment (pshamus) [inline .build/tasks/generateHelp.PlatyPS.build.ps1:111]: This assumes the docs are relative to the module source, not the project (BuildRoot). Most of what I have seen from other projects is that the MD docs are in the root of the project.

File: .build/tasks/generateHelp.PlatyPS.build.ps1 [hunk 3]
```diff
@@ -41,26 +48,162 @@ Task UpdateHelp {
...
+    $ModuleVersionObject = Split-ModuleVersion -ModuleVersion $ModuleVersion
+    $ModuleVersionFolder = $ModuleVersionObject.Version
+    $preReleaseTag       = $ModuleVersionObject.PreReleaseString
+
+    "`tModule Version           = '$ModuleVersion'"
+    "`tModule Version Folder    = '$ModuleVersionFolder'"
+    "`tPre-release Tag          = '$preReleaseTag'"
+
+    $HelpFolder = Get-SamplerAbsolutePath -Path $HelpFolder -RelativeTo $SourcePath
+    "`tHelp Folder Path         = '$HelpFolder'"
```
Comment (pshamus) [inline .build/tasks/generateHelp.PlatyPS.build.ps1:191]: Depending on what is determined in the above comment, this would also need to be modified.

**Review by pshamus — COMMENTED** [review]
Overall looks good. The main issue I have is with the output paths to the help files.

File: .build/tasks/generateHelp.PlatyPS.build.ps1 (hunk 1 above)
Comment (gaelcolas) [reply .build/tasks/generateHelp.PlatyPS.build.ps1:116]: true, I just did not bother with the details for now :)

File: .build/tasks/generateHelp.PlatyPS.build.ps1 (hunk 2 above)
Comment (gaelcolas) [reply .build/tasks/generateHelp.PlatyPS.build.ps1:111]: hmmmmm. Good question, @johlju what do you think?
Technically you could override the default (when I've implemented the config), using, say. `../docs`.

Comment (gaelcolas) [reply .build/tasks/generateHelp.PlatyPS.build.ps1:111]: I'd add that maybe we should have an option to build the MAML into the built module base's CultureInfo folder.

File: .build/tasks/generateHelp.PlatyPS.build.ps1 [hunk 4]
```diff
@@ -14,23 +14,30 @@ param
 
     [Parameter()]
     [System.String]
-    $HelpFolder = (property HelpFolder 'docs'),
+    $HelpFolder = (property HelpFolder 'help'),
```
[the remark this answers is not in the document]
Comment (gaelcolas) [reply .build/tasks/generateHelp.PlatyPS.build.ps1:17]: agreed for the help source.
Where should the MAML be copied though...?
I think we need it within `$BuiltModuleBase\$((Get-Culture).Name)\`, but we may want to have the help also built somewhere else like `$outputDirectory\help\$ModuleVersion\$((Get-Culture).Name)` ?

**Comment by gaelcolas** [comment]
So I've created 3 tasks:
- Generate_maml_from_built_module
- update_markdown_help_source
- Generate_MAML_From_Markdown_help_source

> **_NOTES: Make sure you understand help priorities between Comment based help, Help file, and what you need to change, because this module is not updating the comment-based help for you (i.e. `# .ExternalHelp` entry). That means 'Generate_maml_from_built_module' is probably not really usable yet. (this is still a draft, afterall)_**

> from the doc: _Note that the `.EXTERNALHELP` keyword takes precedence over all other comment-based help keywords. When `.EXTERNALHELP` is present, the `Microsoft.PowerShell.Commands.GetHelpCommand` cmdlet **does not display** comment-based help, even when it cannot find a help file that matches the value of the keyword._

## Generate_maml_from_built_module

This task is used to take the current Comment Based help (CBH) from the module, and generate a MAML help file based on this info.
The process will first convert the CBH to Markdown. Then from Markdown to MAML.
Because you can't check the Markdown or MAML and it's likely to loose some formatting, that's not the best workflow (it's done at build time in the output).
However, that means we can build a MAML and make it available outside of a Module release (say if you branch out a tag just to update the help).

## update_markdown_help_source

This task is used to either generate or update the Markdown help source files (by default in the 'docs/' folder, at the Project's root).

This task can be used during the Dev workflow when a function has had a parameter added.
The task will look for the Markdown file (For every locale that exists under `docs/` and will update it without changing the rest of the document.
Bear in mind that if the parameter description is in English, it will still use the same content for each Locale/CultureInfo.

## Generate_MAML_From_Markdown_help_source

When the source is up to date, it can be used to generate the MAML file (for each locale/CultureInfo).

This will look into `docs/` for the source markdown and create the MAML files in the `output/help` folder, with a version subfolder and locale subfolders:

```
SAMPLER\OUTPUT\HELP
└───0.109.10-preview0001
    ├───en-US
    │       Sampler-help.xml
    │
    └───fr-FR
            Sampler-help.xml
```

Those will also be copied (by default) to the built module folder (i.e. `SAMPLER\OUTPUT\SAMPLER\0.109.10\en-US\Sampler-help.xml`).


I have not yet parameterized some of the options for the PlatyPS commands, hence why it's still a draft.
I'm also looking at how to enable updated help, when we have CBH (i.e. automatically adding `.EXTERNALHELP` at compile time).

Let me know your thoughts.

File: .build/tasks/generateHelp.PlatyPS.build.ps1 (hunk 3 above)
Comment (gaelcolas) [reply .build/tasks/generateHelp.PlatyPS.build.ps1]: should be fixed.

File: .build/tasks/generateHelp.PlatyPS.build.ps1 (hunk 1 above)
Comment (gaelcolas) [reply .build/tasks/generateHelp.PlatyPS.build.ps1]: still on the TODO list.

**Comment by pshamus** [comment]
Other than inconsistent case of the task names, all those sound good. For `Generate_MAML_From_Markdown_help_source`, what's the reasoning behind generating the help in `output\help` rather than just putting it in the built module?

**Comment by gaelcolas** [comment]
generating in `output/help` to make the MAML available (say you want to publish in CAB or make available separately).
There's a copy to Built Module Base by default, so it will be available there anyway too.

We can't directly put it to BuiltModuleBase because then it might be "harder" to get it from there (exact path, is there some other files like *.strings.psd1, maybe we have a Locale folder without MAML...).

**Comment by pshamus** [comment]
Gotcha, makes total sense. Thanks!

**Review by pshamus — COMMENTED** [review]
> Reviewed 23 of 26 files at r1, 2 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/gaelcolas/sampler/260)* status: 25 of 27 files reviewed, 3 unresolved discussions (waiting on @gaelcolas, @johlju, and @pshamus)

___
*[.build/tasks/generateHelp.PlatyPS.build.ps1, line 41 at r3](https://reviewable.io/reviews/gaelcolas/sampler/260#-MW_KjvbEHFVVJwKBPp1:-MW_KjvbEHFVVJwKBPp2:b-v4nb2u) ([raw file](https://github.com/gaelcolas/sampler/blob/e4ffbca789bd180309ba099a43c496a274b68e28/.build/tasks/generateHelp.PlatyPS.build.ps1#L41)):*
> ```PowerShell
>     [Parameter()]
>     [System.Management.Automation.SwitchParameter]
>     $CopyHelpMamlToBuiltModuleBase = (property CopyHelpMamlToBuiltModuleBase $True),
> ```

Not sure if you're a stickler for consistent casing, but this would be `$true`.

**Comment by johlju** [comment]
@gaelcolas 

> I'm also looking at how to enable updated help, when we have CBH (i.e. automatically adding .EXTERNALHELP at compile time).

Could we use ScriptAnalyzer "Fix" functionality by adding a custom rule that supports adding (fixing) this? 🤔 I just saw the other day Chris mentioned that "Fix" triggers a method or function (don't remember), but worth checking out. If I'm not misunderstand what you meant. 🙂

**Review by johlju — COMMENTED** [review]
> Reviewed 24 of 26 files at r1, 2 of 3 files at r2, 2 of 2 files at r4.
> *[Reviewable](https://reviewable.io/reviews/gaelcolas/sampler/260)* status: all files reviewed, 4 unresolved discussions (waiting on @gaelcolas and @pshamus)

*[a discussion](https://reviewable.io/reviews/gaelcolas/sampler/260#-MWmIEPp9DhD3ArOyGsl:-MWmIEPp9DhD3ArOyGsm:b-49l0yj) (no related file):*
Consider update the README.md with the information you provided in the issue comment, or something similar, under section ```## Tasks` (a good start with the information you provided in the issue comment). :slightly_smiling_face: 

---
___
*[.build/tasks/generateHelp.PlatyPS.build.ps1, line 111 at r1](https://reviewable.io/reviews/gaelcolas/sampler/260#-MWTaRr1sWlXuma7JoGs-r1-111:-MWmJRNZ4CKf6g-SeLf6:bdlhynj) ([raw file](https://github.com/gaelcolas/sampler/blob/df82609ba6d1501d36e8bebc770e0311aefc4236/.build/tasks/generateHelp.PlatyPS.build.ps1#L111)):*
<details><summary><i>Previously, gaelcolas (Gael) wrote…</i></summary><blockquote>

I'd add that maybe we should have an option to build the MAML into the built module base's CultureInfo folder.
</blockquote></details>

Normally I think that everything that should trigger the pipeline should be in "source" folder (except for PRs where it always runs), meaning the pipeline will deploy a new preview version on any changes inside "source" folder. We could of course add '/docs' to be included as well. At the same time I think it is starting to be normal for the "code" and "docs" that contributors can contribute to, to be in the source folder. So I would like having the default to be in the source folder and then that a module owner can override it.
___
*[.build/tasks/generateHelp.PlatyPS.build.ps1, line 161 at r4](https://reviewable.io/reviews/gaelcolas/sampler/260#-MWmLoTqCPOQp8OwOgR0:-MWmLoTqCPOQp8OwOgR1:bvltmdp) ([raw file](https://github.com/gaelcolas/sampler/blob/1af78e7016d1dc53393f386db9daad7aba6538fd/.build/tasks/generateHelp.PlatyPS.build.ps1#L161)):*
> ```
> $generateHelpCommands = @"
> ```

Maybe mention why we need to do this so the next contributor know? I think it has to do with different versions of dependent modules?

**Review by johlju — COMMENTED** [review]
:lgtm: 

> *[Reviewable](https://reviewable.io/reviews/gaelcolas/sampler/260#-:-MWmMpNg2R7dkJ-niDOk:bg7igkv)* status: all files reviewed, 4 unresolved discussions (waiting on @gaelcolas and @pshamus)

**Comment by gaelcolas** [comment]
Yeah, I'm not convinced for `source/docs` vs `docs`.
For instance, `tests` is not under `source` yet it should trigger a pipeline if changed (but not a module release).
Same goes for `docs`, maybe we should be able to update the docs/help/maml without the need to release a new module.
And maybe the `docs` folder has more than the commands? Maybe detailed explanation on how to contribute to the project (linked from contributing.md).
In that case you may not want to hide it under source...
And the what do we do with wikisource, is it the same? Different?

**Comment by johlju** [comment]
When tests are update they are verified to run in the PR so they don't need to run when merged to default branch. 

We could say that since docs should not be updated manually it should not be in source, and say that any docs that should be update manually and is published to one or more sources should be in source folder. But there is no reason to merge docs that are updated automatically in the repo. It can be built on each run. 

What is the reason for merging it after it been built?

**Comment by johlju** [comment]
Btw. it looks like the tasks work in my quick test. But why is the docs generated both to /output folder and to /docs folder? 🤔 Isn't output folder enough for it to be published or added to the release?

**Comment by gaelcolas** [comment]
Different tasks for different use case, that's what I explained in the comment above https://github.com/gaelcolas/Sampler/pull/260#issuecomment-805855252

**Review by pshamus — APPROVED** [review]
Looks good to me.

**Comment by gaelcolas** [comment]
I'll merge this when test passes and add configuration option in `build.yaml` in another PR.

**Pull request merged**